File: minimum-number-of-operations-to-make-array-empty/minimum-number-of-operations-to-make-array-empty.java
class Solution {
    public int minOperations(int[] nums) {
        // Create a HashMap to store the frequency of each element in 'nums'
        Map<Integer, Integer> map = new HashMap<>();
        
        // Count the frequency of each element in 'nums'
        for(int num: nums){
            // Put the element in the map and increment its count by 1
            map.put(num, map.getOrDefault(num, 0)+1);
        }
        // Initialize a variable to count the minimum operations needed
        int count = 0;

        // Iterate through the entries in the HashMap
        for(Map.Entry<Integer, Integer> entry: map.entrySet()){
            // Get the frequency count of the current element
            int current = entry.getValue();
            // If there is only one occurrence of any element, it's not possible to make it divisible by 3
            if(current == 1) return -1;

            // Calculate the number of operations needed to make the frequency divisible by 3
            // Count the complete sets of 3 elements
            count += current/3;
            // If there are remaining elements, increment the count
            if(current % 3 != 0)
                count++;
        }
        // Return the total count of operations needed
        return count;
    }
}

File: minimum-number-of-operations-to-make-array-empty/minimum-number-of-operations-to-make-array-empty.cpp
class Solution {
public:
    int minOperations(vector<int>& nums) {
        // Map to store the frequency of each element in the array
        unordered_map<int, int> map;
        
        // Count the frequency of each element in the array
        for (auto i : nums) {
            map[i]++;
        }
        
        // Variable to store the total number of operations
        int count = 0;
        
        // Iterate through the frequency map
        for (auto current : map) {
            // If any element occurs only once, a harmonious array is not possible
            if (current.second == 1) {
                return -1;
            }
            
            // Calculate the number of operations required for each element to be harmonious
            count += current.second / 3;
            
            // If there is a remainder, one additional operation is needed
            if (current.second % 3) {
                count++;
            }
        }
        
        // Return the total number of operations
        return count;
    }
};
